### PR TITLE
Mint Tokens flow wiring

### PR DESCRIPTION
**Pull request by alicjakujawa**
## Description

- This PR wires up the "Mint tokens" flow

**New stuff** ✨

* MintTokens ColonyActions type
* MintTokens ColonyAction event and action copies
* COLONY_ACTION_MINT_TOKENS_* redux action types
* Saga for mintTokens transaction 

**Changes** 🏗
* Refactored TokenMintDialog
* Refactored TokenMintForm
* Updated DetailsWidget to not display from/to domain for MintTokens action type

**Deletions** ⚰️

* ColonyInitialFunding component
* Mint Tokens button from ColonyFundingMenu
* colonyMintTokens saga from admin/sagas
* COLONY_MINT_TOKENS__* types from types/actions/colony.ts

https://user-images.githubusercontent.com/13069555/102783072-af4cc280-439a-11eb-9913-fb744bbc996c.mov

## TODO

- Find out how to deal with Claim tokens transaction (Do we display any information on ActionPage?)
- Refactor Dialogs which are used with wizard and without wizard, to detect when to show "back" button, and when to show "cancel" button 

Resolves DEV-65

### CODE REVIEW

**Comment by rdig** [comment]
@alicjakujawa make sure you re-rebase your branch on the latest changes from `feature/dev64-transfer-funds-flow-wiring`

Let me know if you need help, as it looks like a lot of conflicting files _(although that shouldn't be the case, and Github is just being stupid)_

**Comment by alicjakujawa** [comment]
@rdig  Apparently its github being stupid 😆

File: src/data/resolvers/colonyActions.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ interface EventValue {
   token: Address;
   fromPot: BigNumberish;
   toPot: BigNumberish;
+  who: Address;
```
Comment (rdig) [inline src/data/resolvers/colonyActions.ts:19]: I know this comes from the event, but I so **hate** this prop name!

File: src/modules/dashboard/components/ActionsPage/DetailsWidget/DetailsWidget.tsx [hunk 2]
```diff
@@ -75,6 +75,14 @@ const DetailsWidget = ({
...
   const Symbol = () => values?.tokenSymbol as ReactElement;
 
+  // @TODO We need to find a better way of handling for which actions we want display which fields
+  const showFromDomain =
+    !ColonyActions.MintTokens && values?.fromDomain && showFullDetails;
+  const showToDomain =
+    !ColonyActions.MintTokens &&
+    (values?.toDomain || recipient) &&
+    showFullDetails;
+
```
Comment (rdig) [inline src/modules/dashboard/components/ActionsPage/DetailsWidget/DetailsWidget.tsx:85]: This needs to be refactored in a static mapping as I did for the other stuff that is dependent on different actions or event types.

File: src/modules/dashboard/components/ColonyFundingMenu/ColonyFundingMenu.tsx [hunk 3]
```diff
@@ -66,16 +60,7 @@ const ColonyFundingMenu = ({
...
-  const handleMintTokens = useCallback(() => {
-    const nativeToken =
-      tokens && tokens.find(({ address }) => address === nativeTokenAddress);
-    if (nativeToken) {
-      openTokenMintDialog({
-        nativeToken,
-        colonyAddress,
-      });
-    }
-  }, [colonyAddress, nativeTokenAddress, openTokenMintDialog, tokens]);
```
Comment (rdig) [inline src/modules/dashboard/components/ColonyFundingMenu/ColonyFundingMenu.tsx:78]: Why did you remove this? This still needs to be used here.

File: src/modules/dashboard/components/ManageFundsDialog/ManageFundsDialog.tsx [hunk 4]
```diff
@@ -22,8 +22,8 @@ const MSG = defineMessages({
     id: 'dashboard.ManageFundsDialog.transferFundsDescription',
     defaultMessage: 'Move funds between domais.',
   },
-  moveFundsPermissionsText: {
-    id: 'dashboard.ManageFundsDialog.moveFundsPermissionsText',
+  permissionsListText: {
+    id: 'dashboard.ManageFundsDialog.permissionsListText',
```
Comment (rdig) [inline src/modules/dashboard/components/ManageFundsDialog/ManageFundsDialog.tsx:26]: :+1:

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx [hunk 5]
```diff
@@ -35,79 +39,113 @@ interface FormValues {
...
+  colony: Colony;
 }
 
+type Props = DialogProps &
+  Partial<WizardDialogType<object>> &
+  CustomWizardDialogProps;
+
+const displayName = 'dashboard.TokenMintDialog';
+
 const TokenMintDialog = ({
```
Comment (rdig) [inline src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx:53]: These two files are all mixed up!

`TokenMintDialog` contains the form and `TokenMintDialogForm` contains the dialog.

Please swap the contents of these two files around.

File: src/modules/dashboard/components/TokenMintDialog/TokenMintForm.tsx [hunk 6]
```diff
@@ -65,9 +75,9 @@ const TokenMintForm = ({
         mintAmount: 0,
       }}
       validationSchema={validationSchema}
```
Comment (rdig) [inline src/modules/dashboard/components/TokenMintDialog/TokenMintForm.tsx:77]: Please change the validation schema to:
```
const validationSchema = yup.object().shape({
  justification: yup.string(),
  mintAmount: yup
    .number()
    .required(() => MSG.errorAmountRequired)
    .moreThan(0, () => MSG.errorAmountMin),
});
```

File: src/modules/dashboard/sagas/actions.ts [hunk 7]
```diff
@@ -241,10 +242,108 @@ function* createMoveFundsAction({
...
+
+    // create transactions
+    yield fork(createTransaction, mintTokens.id, {
+      context: ClientType.ColonyClient,
+      methodName: 'mintTokens',
+      identifier: colonyAddress,
+      params: [amount],
+      group: {
+        key: batchKey,
+        id: meta.id,
```
Comment (rdig) [inline src/modules/dashboard/sagas/actions.ts:279]: You have `metaId` already destructured, why are you still using the `meta.id` value from the `meta` object ?

File: src/modules/dashboard/sagas/actions.ts [hunk 8]
```diff
@@ -241,10 +242,108 @@ function* createMoveFundsAction({
...
+      ready: false,
+    });
+    yield fork(createTransaction, claimColonyFunds.id, {
+      context: ClientType.ColonyClient,
+      methodName: 'claimColonyFunds',
+      identifier: colonyAddress,
+      params: [nativeTokenAddress],
+      group: {
+        key: batchKey,
+        id: meta.id,
```
Comment (rdig) [inline src/modules/dashboard/sagas/actions.ts:291]: Same as above

**Review by rdig — CHANGES_REQUESTED** [review]
This PR is mostly good. A very good job for the first time attempting to wire up an action. Very nice @alicjakujawa 

Besides the things I'm pointed out in the code review directly, there's two more things:
- Find out the correct copy for the root permission (from) product, then add it in:
Screenshot from 2020-12-22 15-55-14
- The `FriendlyUsername` core component needs to be refactored so that it can support checking the address that's being passed in that it is either a user (as it does now), or a colony (this is the part that needs to be added in). See here: https://github.com/JoinColony/colonyDapp/blob/e54e6ad7bcf97acb675e2f0e98cf125a64f23920/src/modules/dashboard/components/ActionsPage/ActionsPage.tsx#L273-L288

If you put that in, we'll be able to also display a colony's name, on events like minting: 
Screenshot from 2020-12-22 15-19-42

File: src/data/resolvers/colonyActions.ts (hunk 1 above)
Comment (ArmandoGraterol) [reply src/data/resolvers/colonyActions.ts:19]: Same here, it's not a great name definitely hahaha

**Review by ArmandoGraterol — COMMENTED** [review]
Works great for me! Will re-review it once the changes are made

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 5 above)
Comment (alicjakujawa) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx:53]: Im lost :D I see only one dialog which is parent of TokenMintForm

**Review by ArmandoGraterol — APPROVED** [review]
LGTM :ok_hand:

**Comment by rdig** [comment]
Can I re-review this ?

**Comment by alicjakujawa** [comment]
> Can I re-review this ?

More like answer on my comments/questions here:
https://github.com/JoinColony/colonyDapp/pull/2335#discussion_r547872221

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 5 above)
Comment (rdig) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx:53]: So inside the `TokenMintDialog` component **folder** there are two `jsx` files: 
- `TokenMintDialog`
- `TokenMintForm`

Screenshot from 2021-01-05 14-25-36

The "convention" is that the "Dialog" file contains the "wrapper" component files _(the part that contains the `ActionForm`)_, while the other file, the "Form" contains the actual form and dialog component elements _(the part that has all the `DialogSection`s components)_

You can see and example of this with the `TransferFundsDIalog` component: https://github.com/JoinColony/colonyDapp/tree/feature/dev-65-mint-tokens-flow-wiring/src/modules/dashboard/components/TransferFundsDialog

In your case here, they are reversed, the dialog file contains the dialog sections, while the "form" contains the wrapper.

Just swap the content of these two files around, and it should be all good.

_PS: Let me know after you've made this change so I can re-review this. Thanks_

**Comment by rdig** [comment]
I forgot to mention here, that now that the subgraph integration is in, this also needs to add in the required query for the action and the subgraph changes to support that

File: src/modules/dashboard/components/ActionsPage/DetailsWidget/DetailsWidget.tsx [hunk 9]
```diff
@@ -75,6 +79,19 @@ const DetailsWidget = ({
...
+  ]?.includes(actionType);
+  const showToRecipient = DETAILS_FOR_ACTION[
+    ActionPageDetails.toRecipient
+  ]?.includes(actionType);
+  const showToDomain = DETAILS_FOR_ACTION[ActionPageDetails.toDomain]?.includes(
+    actionType,
+  );
+  const showAmount = DETAILS_FOR_ACTION[ActionPageDetails.amount]?.includes(
+    actionType,
+  );
```
Comment (rdig) [inline src/modules/dashboard/components/ActionsPage/DetailsWidget/DetailsWidget.tsx:93]: In its current form, this is not scalable.

Please refactor this into a helper method similar to those inside `utils/events`, but make it so that it's scalable, so you don't have to change it every time you add new values to the `DETAILS_FOR_ACTION` map.

Something similar to this: https://github.com/JoinColony/colonyDapp/blob/5efc8b7820303b4684b55586102d9aa2931cff24/src/utils/events/index.ts#L100-L107

File: src/modules/dashboard/components/ActionsPage/staticMaps.ts [hunk 10]
```diff
@@ -3,6 +3,13 @@ import { ColonyAndExtensionsEvents, ColonyActions } from '~types/index';
 
 import { STATUS } from './types';
 
+export enum ActionPageDetails {
+  fromDomain = 'fromDomain',
+  toDomain = 'toDomain',
+  toRecipient = 'toRecipient',
+  amount = 'amount',
+}
+
```
Comment (rdig) [inline src/modules/dashboard/components/ActionsPage/staticMaps.ts:12]: Please rename this to start with an uppercase letter in order to keep naming consistent.

Eg:
```
FromDomain = 'FromDomain',
```

File: src/modules/dashboard/components/ActionsPage/staticMaps.ts [hunk 11]
```diff
@@ -71,4 +87,23 @@ export const EVENTS_REQUIRED_FOR_ACTION: ActionsEventsMap = {
...
+    ColonyActions.Payment,
+    ColonyActions.MoveFunds,
+  ],
+  [ActionPageDetails.toDomain]: [ColonyActions.MoveFunds],
+  [ActionPageDetails.toRecipient]: [ColonyActions.Payment],
+  [ActionPageDetails.amount]: [
+    ColonyActions.Payment,
+    ColonyActions.MoveFunds,
+    ColonyActions.MintTokens,
+  ],
```
Comment (rdig) [inline src/modules/dashboard/components/ActionsPage/staticMaps.ts:108]: This should be reversed. The details should based on actions, **not** the actions on the details.

Eg:
```
[ColonyActions.Payment]: [
  ActionPageDetails.fromDomain,
  ActionPageDetails.toRecipient,
  ActionPageDetails.amount
],
```

File: src/modules/dashboard/components/ManageFundsDialog/ManageFundsDialog.tsx [hunk 12]
```diff
@@ -107,6 +114,14 @@ const ManageFundsDialog = ({
...
       description: MSG.mintTokensDescription,
       icon: 'emoji-seed-sprout',
+      permissionRequired: !canMintNativeToken,
+      permissionInfoText: MSG.permissionsListText,
+      permissionInfoTextValues: {
+        permissionsList: (
+          <FormattedMessage {...MSG.mintTokensPermissionsList} />
+        ),
+      },
+      onClick: () => callStep(nextStepMintTokens),
```
Comment (rdig) [inline src/modules/dashboard/components/ManageFundsDialog/ManageFundsDialog.tsx:124]: Currently we have a bug here that breaks the `canMintNativeToken` logic

Screenshot from 2021-01-06 19-03-00

It's to do with the Apollo cache and our `ethereal` wallet.

Please open an issue to track this, as it needs to be fixed prior to launch

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx [hunk 13]
```diff
@@ -1,113 +1,118 @@
...
+}: Props) => {
+  const history = useHistory();
+
+  const nativeToken =
+    tokens && tokens.find(({ address }) => address === nativeTokenAddress);
+
+  const transform = useCallback(
+    pipe(
+      mapPayload(({ mintAmount: inputAmount }) => {
+        // Find the selected token's decimals
```
Comment (rdig) [inline src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx:70]: This comment is wrong. You don't **find** them, you're just converting the value to `wei`

File: src/modules/dashboard/components/ActionsPage/staticMaps.ts [hunk 14]
```diff
@@ -24,6 +37,7 @@ export const EVENT_ROLES_MAP: EventRolesMap = {
   [ColonyAndExtensionsEvents.ColonyFundsMovedBetweenFundingPots]: [
     ColonyRole.Funding,
   ],
+  [ColonyAndExtensionsEvents.TokensMinted]: [ColonyRole.Root],
```
Comment (rdig) [inline src/modules/dashboard/components/ActionsPage/staticMaps.ts:40]: Please add the correct message descriptor for this message role type (inside the `ActionsPageEvent` component)

Screenshot from 2021-01-06 19-32-30

Also, the generic copy has a typo in it, please fix it.

File: src/i18n/en-events.ts [hunk 15]
```diff
@@ -6,6 +6,7 @@ const eventsMessageDescriptors = {
   'event.title': `{eventName, select,
       ${ColonyAndExtensionsEvents.OneTxPaymentMade} {{initiator} paid {amount} {tokenSymbol} from {fromDomain} to {recipient}}
       ${ColonyAndExtensionsEvents.ColonyFundsMovedBetweenFundingPots} {{initiator} transferred {amount} {tokenSymbol} from the {fromDomain} to the {toDomain}}
+      ${ColonyAndExtensionsEvents.TokensMinted} {{initiator} minted {amount} {tokenSymbol} to {recipient}}
```
Comment (rdig) [inline src/i18n/en-events.ts:9]: This should show the colony name (in place of the recipient) if one exists, the same thing we do for the user.

So you'll need to create a similar component to my `FriendlyUserName` _(check `ActionsPage` to see how that works)_

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx [hunk 16]
```diff
@@ -1,113 +1,118 @@
...
+      }),
+      withMeta({ history }),
+    ),
+    [],
+  );
+
+  return (
+    <ActionForm
+      initialValues={{
+        justification: '',
```
Comment (rdig) [inline src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx:92]: `annotation`

Please rename it, to keep it consistent throughout the dapp

File: src/modules/dashboard/sagas/actions.ts [hunk 17]
```diff
@@ -241,10 +242,108 @@ function* createMoveFundsAction({
...
+        key: batchKey,
+        id: metaId,
+        index: 1,
+      },
+      ready: false,
+    });
+
+    yield takeFrom(mintTokens.channel, ActionTypes.TRANSACTION_CREATED);
+    yield takeFrom(claimColonyFunds.channel, ActionTypes.TRANSACTION_CREATED);
+    yield put({ type: ActionTypes.COLONY_ACTION_MINT_TOKENS_SUBMITTED });
```
Comment (rdig) [inline src/modules/dashboard/sagas/actions.ts:299]: This is not wired anywhere. What does it do ?

Comment (rdig) [reply src/modules/dashboard/sagas/actions.ts:299]: Right. I found out what it used to do. 

In the previous iteration, it was used to close the modal before the gas station opened. We're not doing that now, so this is pointless.

Please remove this as well as the redux action.

But in the future please investigate better the code you are copy-pasting or refactoring.

File: src/redux/actionTypes.ts [hunk 18]
```diff
@@ -44,6 +40,10 @@ export enum ActionTypes {
   COLONY_ACTION_EXPENDITURE_PAYMENT = 'COLONY_ACTION_EXPENDITURE_PAYMENT',
   COLONY_ACTION_EXPENDITURE_PAYMENT_ERROR = 'COLONY_ACTION_EXPENDITURE_PAYMENT_ERROR',
   COLONY_ACTION_EXPENDITURE_PAYMENT_SUCCESS = 'COLONY_ACTION_EXPENDITURE_PAYMENT_SUCCESS',
+  COLONY_ACTION_MINT_TOKENS = 'COLONY_ACTION_MINT_TOKENS',
+  COLONY_ACTION_MINT_TOKENS_ERROR = 'COLONY_ACTION_MINT_TOKENS_ERROR',
+  COLONY_ACTION_MINT_TOKENS_SUCCESS = 'COLONY_ACTION_MINT_TOKENS_SUCCESS',
+  COLONY_ACTION_MINT_TOKENS_SUBMITTED = 'COLONY_ACTION_MINT_TOKENS_SUBMITTED',
```
Comment (rdig) [inline src/redux/actionTypes.ts:46]: As I pointed out above, this is not needed

File: src/redux/types/actions/colonyActions.ts [hunk 19]
```diff
@@ -55,4 +55,24 @@ export type ColonyActionsActionTypes =
...
+        amount: BigNumber;
+      },
+      MetaWithHistory<object>
+    >
+  | ErrorActionType<ActionTypes.COLONY_ACTION_MINT_TOKENS_ERROR, object>
+  | UniqueActionType<
+      ActionTypes.COLONY_ACTION_MINT_TOKENS_SUBMITTED,
+      object,
+      object
+    >
```
Comment (rdig) [inline src/redux/types/actions/colonyActions.ts:74]: Not needed either

**Review by rdig — CHANGES_REQUESTED** [review]
Functionally this is working fine, however, the code still needs work as a lot of things can be improved, refactored, or removed.

I've pointed out below a bunch of stuff that needs to change.

**Comment by alicjakujawa** [comment]
@rdig Ready for your 👀

File: src/modules/core/components/FriendlyName/FriendlyName.tsx [hunk 20]
```diff
@@ -33,6 +33,10 @@ const FriendlyUserName = ({
 }: Props) => {
   const addressRef = useRef<HTMLElement>(null);
 
+  const { data: colonyData } = useColonyQuery({
+    variables: { address: walletAddress || '' },
+  });
```
Comment (rdig) [inline src/modules/core/components/FriendlyName/FriendlyName.tsx:38]: Please don't fetch data in this component, it's pointless. Just pass it along the same way we do the user profile.

That way you can still keep your display logic intact.

File: src/modules/core/components/FriendlyName/FriendlyName.tsx [hunk 21]
```diff
@@ -57,17 +61,19 @@ const FriendlyUserName = ({
...
-      )}
+      {userDisplayName ||
+        (username && `@${username}`) ||
+        colonyData?.colony.colonyName || (
+          <MaskedAddress
+            address={walletAddress}
+            full={!maskedAddress}
+            ref={addressRef}
+          />
+        )}
```
Comment (rdig) [inline src/modules/core/components/FriendlyName/FriendlyName.tsx:72]: Please make sure to also account for the colony displayName (the same way we account for the userDisplayName)

Here's an example where we have a defined colony name:

Screenshot from 2021-01-12 11-48-47

Yet the `FriendlyName` component just shows the ens name:

Screenshot from 2021-01-12 11-48-39

File: src/modules/dashboard/components/ActionsPage/ActionsPage.tsx [hunk 22]
```diff
@@ -272,18 +272,12 @@ const ActionsPage = () => {
...
+        <FriendlyName user={initiatorProfileWithFallback} autoShrinkAddress />
       </span>
     ),
     recipient: (
       <span className={styles.titleDecoration}>
-        <FriendlyUserName
-          user={recipientProfileWithFallback}
-          autoShrinkAddress
-        />
+        <FriendlyName user={recipientProfileWithFallback} autoShrinkAddress />
```
Comment (rdig) [inline src/modules/dashboard/components/ActionsPage/ActionsPage.tsx:280]: As stated above, this will have to change to pass in the colony details as well

File: src/utils/colonyActions.ts [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+type DetailsValuesMap = Partial<
+  {
+    [key in ActionPageDetails]: boolean;
+  }
+>;
+
+/*
+ * Get colony action details for DetailsWidget based on action type and ActionPageDetails map
+ */
+export const getDetailsForAction = (
```
Comment (rdig) [inline src/utils/colonyActions.ts:17]: Very nice! :+1:

**Review by rdig — APPROVED** [review]
I'm all good with this, except the `FriendlyName` component needs a bit of work (just some minor refactor) -- see comments below.

After that is done, you can merge this, so I'm pre-approving this.

Nice work on your first wiring PR! :100: :1st_place_medal:
